### ISSUE TITLE
Build failure with dynamodb-enhanced extension and async client

### ISSUE THREAD

**Issue by suilven** [quarkiverse/quarkus-amazon-services#458]
Hello,

I've been playing around with the [Quarkus Amazon DynamoDB quickstart](https://github.com/quarkusio/quarkus-quickstarts/tree/main/amazon-dynamodb-quickstart) and I've uncovered what looks like a bug when using an asynchronous dynamo client and the dynamodb-enhanced extension.

I've distilled the problem down to the minimum set of changes to highlight the problem. The diff of my changes is:
```
Index: amazon-dynamodb-quickstart/src/main/java/org/acme/dynamodb/FruitResource.java
===================================================================
diff --git a/amazon-dynamodb-quickstart/src/main/java/org/acme/dynamodb/FruitResource.java b/amazon-dynamodb-quickstart/src/main/java/org/acme/dynamodb/FruitResource.java
deleted file mode 100644
--- a/amazon-dynamodb-quickstart/src/main/java/org/acme/dynamodb/FruitResource.java	(revision 20053c0a201388e02a45120102cddc12c983ccda)
+++ /dev/null	(revision 20053c0a201388e02a45120102cddc12c983ccda)
@@ -1,37 +0,0 @@
-package org.acme.dynamodb;
-
-import java.util.List;
-
-import javax.inject.Inject;
-import javax.ws.rs.Consumes;
-import javax.ws.rs.GET;
-import javax.ws.rs.POST;
-import javax.ws.rs.Path;
-import javax.ws.rs.Produces;
-import javax.ws.rs.core.MediaType;
-
-@Path("/fruits")
-@Produces(MediaType.APPLICATION_JSON)
-@Consumes(MediaType.APPLICATION_JSON)
-public class FruitResource {
-
-    @Inject
-    FruitSyncService service;
-
-    @GET
-    public List<Fruit> getAll() {
-        return service.findAll();
-    }
-
-    @GET
-    @Path("{name}")
-    public Fruit getSingle(String name) {
-        return service.get(name);
-    }
-
-    @POST
-    public List<Fruit> add(Fruit fruit) {
-        service.add(fruit);
-        return getAll();
-    }
-}
\ No newline at end of file
Index: amazon-dynamodb-quickstart/src/main/java/org/acme/dynamodb/FruitSyncService.java
===================================================================
diff --git a/amazon-dynamodb-quickstart/src/main/java/org/acme/dynamodb/FruitSyncService.java b/amazon-dynamodb-quickstart/src/main/java/org/acme/dynamodb/FruitSyncService.java
deleted file mode 100644
--- a/amazon-dynamodb-quickstart/src/main/java/org/acme/dynamodb/FruitSyncService.java	(revision 20053c0a201388e02a45120102cddc12c983ccda)
+++ /dev/null	(revision 20053c0a201388e02a45120102cddc12c983ccda)
@@ -1,31 +0,0 @@
-package org.acme.dynamodb;
-
-import java.util.List;
-import java.util.stream.Collectors;
-
-import javax.enterprise.context.ApplicationScoped;
-import javax.inject.Inject;
-
-import software.amazon.awssdk.services.dynamodb.DynamoDbClient;
-
-@ApplicationScoped
-public class FruitSyncService extends AbstractService {
-
-    @Inject
-    DynamoDbClient dynamoDB;
-
-    public List<Fruit> findAll() {
-        return dynamoDB.scanPaginator(scanRequest()).items().stream()
-                .map(Fruit::from)
-                .collect(Collectors.toList());
-    }
-
-    public List<Fruit> add(Fruit fruit) {
-        dynamoDB.putItem(putRequest(fruit));
-        return findAll();
-    }
-
-    public Fruit get(String name) {
-        return Fruit.from(dynamoDB.getItem(getRequest(name)).item());
-    }
-}
\ No newline at end of file
Index: amazon-dynamodb-quickstart/src/test/java/org/acme/dynamodb/DynamodbResourcesTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/amazon-dynamodb-quickstart/src/test/java/org/acme/dynamodb/DynamodbResourcesTest.java b/amazon-dynamodb-quickstart/src/test/java/org/acme/dynamodb/DynamodbResourcesTest.java
--- a/amazon-dynamodb-quickstart/src/test/java/org/acme/dynamodb/DynamodbResourcesTest.java	(revision 20053c0a201388e02a45120102cddc12c983ccda)
+++ b/amazon-dynamodb-quickstart/src/test/java/org/acme/dynamodb/DynamodbResourcesTest.java	(date 1673022219352)
@@ -22,7 +22,7 @@
         .format("{\"name\":\"%s\", \"description\":\"%s\"}", name, description);
 
     @ParameterizedTest
-    @ValueSource(strings = {"fruits", "async-fruits"})
+    @ValueSource(strings = {"async-fruits"})
     void testResource(final String testedResource) {
         List<String> data = Arrays.asList("Cherry", "Pear");
 
Index: amazon-dynamodb-quickstart/pom.xml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/amazon-dynamodb-quickstart/pom.xml b/amazon-dynamodb-quickstart/pom.xml
--- a/amazon-dynamodb-quickstart/pom.xml	(revision 20053c0a201388e02a45120102cddc12c983ccda)
+++ b/amazon-dynamodb-quickstart/pom.xml	(date 1673022281173)
@@ -50,11 +50,12 @@
         </dependency>
         <dependency>
             <groupId>io.quarkiverse.amazonservices</groupId>
-            <artifactId>quarkus-amazon-dynamodb</artifactId>
+            <artifactId>quarkus-amazon-dynamodb-enhanced</artifactId>
         </dependency>
         <dependency>
             <groupId>software.amazon.awssdk</groupId>
             <artifactId>url-connection-client</artifactId>
+            <scope>test</scope>
         </dependency>
         <dependency>
             <groupId>software.amazon.awssdk</groupId>
```
Essentially, I've removed the synchronous code and swapped the `quarkus-amazon-dynamodb` extension for the `quarkus-amazon-dynamodb-enhanced` extension. The Netty client was already included in the dependencies and I've moved the `url-connection-client` to the `test` scope because its still required by `DynamodbResource` in order to create a DynamoDB table in localstack to run the tests. When I try to package the build, it fails with the following stacktrace:
```
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal io.quarkus:quarkus-maven-plugin:2.15.0.Final:build (default) on project amazon-dynamodb-quickstart: Failed to build quarkus application
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2 (MojoExecutor.java:375)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute (MojoExecutor.java:351)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:215)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:171)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:163)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:294)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:960)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:293)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:196)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
    at org.codehaus.classworlds.Launcher.main (Launcher.java:47)
Caused by: org.apache.maven.plugin.MojoExecutionException: Failed to build quarkus application
    at io.quarkus.maven.BuildMojo.doExecute (BuildMojo.java:170)
    at io.quarkus.maven.QuarkusBootstrapMojo.execute (QuarkusBootstrapMojo.java:154)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2 (MojoExecutor.java:370)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute (MojoExecutor.java:351)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:215)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:171)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:163)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:294)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:960)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:293)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:196)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
    at org.codehaus.classworlds.Launcher.main (Launcher.java:47)
Caused by: java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
	[error]: Build step io.quarkus.amazon.common.deployment.AmazonServicesClientsProcessor#setup threw an exception: javax.enterprise.inject.spi.DeploymentException: Missing 'software.amazon.awssdk:url-connection-client' dependency on the classpath
	at io.quarkus.amazon.common.deployment.AmazonServicesClientsProcessor.missingDependencyException(AmazonServicesClientsProcessor.java:173)
	at io.quarkus.amazon.common.deployment.AmazonServicesClientsProcessor.setup(AmazonServicesClientsProcessor.java:122)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:909)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:281)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
	at java.base/java.lang.Thread.run(Thread.java:829)
	at org.jboss.threads.JBossThread.run(JBossThread.java:501)

    at io.quarkus.runner.bootstrap.AugmentActionImpl.runAugment (AugmentActionImpl.java:335)
    at io.quarkus.runner.bootstrap.AugmentActionImpl.createProductionApplication (AugmentActionImpl.java:175)
    at io.quarkus.maven.BuildMojo.doExecute (BuildMojo.java:133)
    at io.quarkus.maven.QuarkusBootstrapMojo.execute (QuarkusBootstrapMojo.java:154)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2 (MojoExecutor.java:370)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute (MojoExecutor.java:351)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:215)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:171)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:163)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:294)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:960)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:293)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:196)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
    at org.codehaus.classworlds.Launcher.main (Launcher.java:47)
Caused by: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
	[error]: Build step io.quarkus.amazon.common.deployment.AmazonServicesClientsProcessor#setup threw an exception: javax.enterprise.inject.spi.DeploymentException: Missing 'software.amazon.awssdk:url-connection-client' dependency on the classpath
	at io.quarkus.amazon.common.deployment.AmazonServicesClientsProcessor.missingDependencyException(AmazonServicesClientsProcessor.java:173)
	at io.quarkus.amazon.common.deployment.AmazonServicesClientsProcessor.setup(AmazonServicesClientsProcessor.java:122)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:909)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:281)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
	at java.base/java.lang.Thread.run(Thread.java:829)
	at org.jboss.threads.JBossThread.run(JBossThread.java:501)

    at io.quarkus.builder.Execution.run (Execution.java:123)
    at io.quarkus.builder.BuildExecutionBuilder.execute (BuildExecutionBuilder.java:79)
    at io.quarkus.deployment.QuarkusAugmentor.run (QuarkusAugmentor.java:160)
    at io.quarkus.runner.bootstrap.AugmentActionImpl.runAugment (AugmentActionImpl.java:331)
    at io.quarkus.runner.bootstrap.AugmentActionImpl.createProductionApplication (AugmentActionImpl.java:175)
    at io.quarkus.maven.BuildMojo.doExecute (BuildMojo.java:133)
    at io.quarkus.maven.QuarkusBootstrapMojo.execute (QuarkusBootstrapMojo.java:154)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2 (MojoExecutor.java:370)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute (MojoExecutor.java:351)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:215)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:171)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:163)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:294)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:960)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:293)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:196)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
    at org.codehaus.classworlds.Launcher.main (Launcher.java:47)
Caused by: javax.enterprise.inject.spi.DeploymentException: Missing 'software.amazon.awssdk:url-connection-client' dependency on the classpath
    at io.quarkus.amazon.common.deployment.AmazonServicesClientsProcessor.missingDependencyException (AmazonServicesClientsProcessor.java:173)
    at io.quarkus.amazon.common.deployment.AmazonServicesClientsProcessor.setup (AmazonServicesClientsProcessor.java:122)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at io.quarkus.deployment.ExtensionLoader$3.execute (ExtensionLoader.java:909)
    at io.quarkus.builder.BuildContext.run (BuildContext.java:281)
    at org.jboss.threads.ContextHandler$1.runWith (ContextHandler.java:18)
    at org.jboss.threads.EnhancedQueueExecutor$Task.run (EnhancedQueueExecutor.java:2449)
    at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run (EnhancedQueueExecutor.java:1478)
    at java.lang.Thread.run (Thread.java:829)
    at org.jboss.threads.JBossThread.run (JBossThread.java:501)
```
If I revert to the `quarkus-amazon-dynamodb` extension but keep my other changes, the build succeeds.

I've also tried adding the Apache HTTP client to the compile classpath but the build fails with the same error, i.e. the enhanced extension seems to require the `url-connection-client` to be on the compile classpath.

**Comment by hamburml** [comment]
Hi!

> Essentially, I've removed the synchronous code and swapped the quarkus-amazon-dynamodb extension for the quarkus-amazon-dynamodb-enhanced extension. 

You need quarkus-amazon-dynamodb and quarkus-amazon-dynamodb-enhanced extension. Pls check this PR https://github.com/quarkusio/quarkus-quickstarts/pull/1186 which is sadly not merged into quickstart repo.

**Comment by suilven** [comment]
Thanks @hamburml

I tried your suggestion and added both dynamodb extensions, but it makes no difference. Unless I have `url-connection-client` on the compile classpath, the build complains about the missing dependency. With just the `quarkus-amazon-dynamodb` extension the build succeeds without  `url-connection-client` on the compile classpath - presumably because the Netty client is present. This is exactly the behaviour I want, but I also want to use the enhanced DynamoDB client. Something about the enhanced extension seems to interfere with the process of finding a suitable client on the classpath, but I don't understand enough about Quarkus extensions to work out what the problem might be.

**Comment by hamburml** [comment]
This repo has tests with async and sync examples. This is the location of the pom https://github.com/quarkiverse/quarkus-amazon-services/blob/main/integration-tests/pom.xml. It has `url-connection-client` and `netty-nio-client` as dependencies. You could compare it to your pom (or share it here if you like).

Here is the EnhancedResource https://github.com/quarkiverse/quarkus-amazon-services/blob/main/integration-tests/src/main/java/io/quarkus/it/amazon/dynamodb/enhanced/DynamoDbEnhancedResource.java with async usage.

I think you will always need `url-connection-client`. And if you want async you will always additionally need `netty-nio-client` https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/asynchronous.html#advanced-operations

>  I've moved the url-connection-client to the test scope ...

I am not 100 percent sure but if you set this to test scope it is not available anymore during compile.

**Comment by suilven** [comment]
Thanks again for your comments.

Unfortunately, I still don't understand why I'm forced to add `url-connection-client` to the classpath if I only want to use an async client? I'm writing a Lambda which accesses DynamoDB, so minimising dependencies is important to me. If I use the `quarkus-amazon-dynamodb` extension with as async DynamoDB client in my code, Quarkus builds my application successfully if I have the `netty-nio-client` on the classpath but not `url-connection-client`. As soon as I switch to the `quarkus-amazon-dynamodb-enhanced` extension, I have to add `url-connection-client` to the classpath, even if I have only async DynamoDB clients in my code.

Maybe this is expected behaviour but I'm fairly certain if I used the AWS SDK directly (without the Quarkus extension) I could build the enhanced async DynamoDB client with only the Netty HTTP client on the classpath. I will try that out and let you know.

**Comment by hamburml** [comment]
Sorry, now I got it :)

The exception should be thrown here https://github.com/quarkiverse/quarkus-amazon-services/blob/main/common/deployment/src/main/java/io/quarkus/amazon/common/deployment/AmazonServicesClientsProcessor.java#L122

`        boolean syncTransportNeeded = amazonClients.stream().anyMatch(item -> item.getSyncClassName().isPresent());`

Can you provide a short project as example?

**Comment by suilven** [comment]
Thanks @hamburml 
I've been looking at that code too, but I don't know enough about CDI and Quarkus extensions to understand what's going wrong.

I've added a short project at https://github.com/suilven/dynamodb-enhanced-dependencies to highlight the problem. This project does nothing useful but clearly demonstrates the build failure.

**Comment by hamburml** [comment]
Looks like https://github.com/quarkiverse/quarkus-amazon-services/blob/main/dynamodb-enhanced/runtime/src/main/java/io/quarkus/amazon/dynamodb/enhanced/runtime/DynamodbEnhancedClientProducer.java#L20 is the issue. 

```
    DynamodbEnhancedClientProducer(Instance<DynamoDbClient> syncClientInstance,
            Instance<DynamoDbAsyncClient> asyncClientInstance) {
```
always creates a DynamoDbClient and AsyncClient, therefore it thinks that url-connection-client is necessary. I build the extension locally with 999-SNAPSHOT version, added some logging and tried it out with your example (and switched there also to 999-SNAPSHOT).

I will try some stuff tomorrow evening :)

**Comment by suilven** [comment]
Thank-you @hamburml :smile: 
I will be interested to see the fix for this one and hopefully learn a little more about how Quarkus extensions work.
Let me know if you need any input from me.

**Comment by hamburml** [comment]
Hm... That seems trickier than I thought it would.

I tried two separate ClientProducers (one syncClientProducer, one asyncClientProducer) and registered both like here https://github.com/quarkiverse/quarkus-amazon-services/blob/main/dynamodb-enhanced/deployment/src/main/java/io/quarkus/amazon/dynamodb/enhanced/deployment/DynamodbEnhancedProcessor.java#L42 - did not work.

I tried it with their builder (DynamoDbEnhancedClient.Builder and DynamoDbEnhancedAsyncClient.Builder) which did not work (like here https://github.com/quarkiverse/quarkus-amazon-services/blob/main/dynamodb/runtime/src/main/java/io/quarkus/amazon/dynamodb/runtime/DynamodbClientProducer.java#L20) but the enhancedClient builders work differently.

And I tried it with an empty Constructor which creates the Clients during runtime.

```
    DynamodbEnhancedClientProducer() {

        this.syncEnhancedClient = DynamoDbEnhancedClient.builder()
                .dynamoDbClient(Arc.container().instance(DynamoDbClient.class).get()).build();

        this.asyncEnhancedClient = DynamoDbEnhancedAsyncClient.builder()
                .dynamoDbClient(Arc.container().instance(DynamoDbAsyncClient.class).get()).build();

    }
```
unfortunately than the checks here https://github.com/quarkiverse/quarkus-amazon-services/blob/main/common/deployment/src/main/java/io/quarkus/amazon/common/deployment/AmazonServicesClientsProcessor.java#L99 dont work anymore because during buildtime the amazonClients are not instantiated, they are only instantiated during runtime....

@scrocquesel Do you have an idea? In the PR where we added the EnhancedClient you helped with Instance<...>. Is there a better way which does not always create both clients even when only one (async for example) is necessary?

**Comment by scrocquesel** [comment]
The issue is that the extension ensure usage of the client will be honored. It does so by looking for injection point.

https://github.com/quarkiverse/quarkus-amazon-services/blob/97bc0efee34941d2936e50f12b49639352cc0abc/common/deployment/src/main/java/io/quarkus/amazon/common/deployment/AbstractAmazonServiceProcessor.java#L65-L78

The enhanced producer make use of `injectionPoint.isProgrammaticLookup()` so even if the sync client is not used, the enhanced extension declare it.
I'm not sure we can simply exclude programmatic lookup without breaking change.

I'm no expert, but, you may reuse the client builder like the dynamodb producer. Client instance will not be shared but that should not be an issue.
Or you can try to lazily set the client instance on the producer with a `BeanContainerListenerBuildItem` and a recorder.
In both case, the extension will need to require the dynamodb client to be produced, by discovering injection point of the enhanced version for project that only use the enhanced client.

I have quickly pushed something that build the sample gradle project but didn't test it

https://github.com/scrocquesel/quarkus-amazon-services/tree/dynamodb

**Comment by scrocquesel** [comment]
@hamburml @suilven Can you give a try to the PR ?

**Comment by hamburml** [comment]
I will, just need some free time. Maybe in the evening. Thanks for your help and your explanation.

**Comment by suilven** [comment]
@scrocquesel thanks for your PR. I've tested it out and I can confirm it works :tada:

Using my example project, I no longer need to declare a dependency on `url-connection-client` with the `DynamoDbEnhancedAsyncClient`. If I switch to the `DynamoDbEnhancedClient` I do need to decalre `url-connection-client` as a dependency. I notice `netty-nio-client` is pulled in for both clients, but that's easily excluded if have only a sync client.

From my perspective, this fixes the issue I reported :smiley:
Thank-you for your help.
